pkg/services/ngalert/state/manager.go

```diff
@@ -297,6 +307,42 @@ func (st *Manager) Put(states []*State) {
 	}
 }
 
+// flushState dumps the entire state to the database
+func (st *Manager) flushState(ctx context.Context) {
+	t := st.clock.Now()
+	st.log.Info("flushing the state")
+	st.cache.mtxStates.Lock()
+	defer st.cache.mtxStates.Unlock()
+	totalStates, errorsCnt := 0, 0
+	for _, orgStates := range st.cache.states {
+		for _, ruleStates := range orgStates {
+			for _, state := range ruleStates {
+				err := st.saveState(ctx, state)
+				totalStates++
+				if err != nil {
+					st.log.Error("failed to save alert state", append(state.GetRuleKey().LogContext(), "labels", state.Labels.String(), "state", state.State.String(), "err", err.Error()))
```

Comment:
What does it mean to do a partial flush of the state? Should we not return err here instead?

Reply:
Good question that I asked myself as well. I moved the code as is. I think the idea was to save as many states as possible. The error here can come from instance validation or SQL.  I think skipping the individual state was motivated by the former error source.